### PR TITLE
Fix single player puzzle update

### PR DESCRIPTION
**Pull request by digit01Wave**
#973 

Fixed issue when switching a puzzle between IsForSingleUser and a regular team puzzle.

### CODE REVIEW

File: ServerCore/Pages/Puzzles/Edit.cshtml.cs [hunk 1]
```diff
@@ -120,6 +120,41 @@ public async Task<IActionResult> OnPostAsync(int puzzleId)
...
+                        _context.HintStatePerTeam.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<PuzzleStatePerTeam> puzzleStatesToRemove = _context.PuzzleStatePerTeam.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.PuzzleStatePerTeam.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<Submission> submissionsToRemove = _context.Submissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.Submissions.RemoveRange(submissionsToRemove);
+                    }
+
+                    await _context.SaveChangesAsync();
```
Comment (jbodner09) [inline ServerCore/Pages/Puzzles/Edit.cshtml.cs:155]: Is there any reason this new block shouldn't just come before the existing SaveChangesAsync, removing the need for a second database save?

File: ServerCore/Pages/Puzzles/Edit.cshtml.cs [hunk 2]
```diff
@@ -120,6 +120,41 @@ public async Task<IActionResult> OnPostAsync(int puzzleId)
...
+                        IEnumerable<HintStatePerTeam> hintStatesToRemove = _context.HintStatePerTeam.Where(hintState => hintState.Hint.PuzzleID == oldPuzzleView.ID);
+                        _context.HintStatePerTeam.RemoveRange(hintStatesToRemove);
+
+                        IEnumerable<PuzzleStatePerTeam> puzzleStatesToRemove = _context.PuzzleStatePerTeam.Where(puzzleState => puzzleState.PuzzleID == oldPuzzleView.ID);
+                        _context.PuzzleStatePerTeam.RemoveRange(puzzleStatesToRemove);
+
+                        IEnumerable<Submission> submissionsToRemove = _context.Submissions.Where(submissions => submissions.PuzzleID == oldPuzzleView.ID);
+                        _context.Submissions.RemoveRange(submissionsToRemove);
+                    }
+
```
Comment (jbodner09) [inline ServerCore/Pages/Puzzles/Edit.cshtml.cs:154]: We also need to create the appropriate rows in the database for the new puzzle type.  If it's now a single player puzzle, SinglePlayerPuzzleUnlockStates needs to be created, and if it's now a team puzzle, PuzzleStatePerTeam needs to be created for every team.  Mirroring the logic in https://github.com/PuzzleServer/mainpuzzleserver/blob/main/ServerCore/Pages/Puzzles/Create.cshtml.cs#L90-L103 should be sufficient.

**Review by jbodner09 — CHANGES_REQUESTED** [review]

File: ServerCore/Pages/Puzzles/Edit.cshtml.cs (hunk 2 above)
Comment (tabascq) [reply ServerCore/Pages/Puzzles/Edit.cshtml.cs:154]: +1

Comment (tabascq) [reply ServerCore/Pages/Puzzles/Edit.cshtml.cs:154]: Or is this what new lines 218-228 are doing?

File: ServerCore/Pages/Puzzles/Edit.cshtml.cs (hunk 1 above)
Comment (tabascq) [reply ServerCore/Pages/Puzzles/Edit.cshtml.cs:155]: Unless there's a reason why the new code should come first (and please document why if so), I'd just use the same if/then for both blocks of code (old line 163, new line 198) for a) maximum clarity and b) full atomicity of the DB change.

Comment (digit01Wave) [reply ServerCore/Pages/Puzzles/Edit.cshtml.cs]: Done!

File: ServerCore/Pages/Puzzles/Edit.cshtml.cs (hunk 2 above)
Comment (digit01Wave) [reply ServerCore/Pages/Puzzles/Edit.cshtml.cs]: It is what those lines were doing

**Review by tabascq — APPROVED** [review]

**Review by jbodner09 — APPROVED** [review]

**Pull request merged**